### PR TITLE
fix: partitioned semaphore resolver

### PR DESCRIPTION
**Pull request by grutt**
# Description

Improvements for handling semaphores in invalid states

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: internal/services/ticker/ticker.go [hunk 1]
```diff
@@ -115,6 +128,10 @@ func New(fs ...TickerOpt) (*TickerImpl, error) {
 		return nil, fmt.Errorf("tenant alerter is required. use WithTenantAlerter")
 	}
 
+	if opts.partitionId == "" {
```
Comment (abelanger5) [inline internal/services/ticker/ticker.go:131]: You're not setting this opt in the `TickerImpl`

File: internal/services/ticker/ticker.go [hunk 2]
```diff
@@ -350,21 +367,72 @@ func (t *TickerImpl) runStreamEventCleanup(ctx context.Context) func() {
...
+		}
+
+		n, err := t.repo.Worker().ResolveWorkerSemaphoreSlots(ctx, tenantId)
+
+		if n.HasResolved {
+			t.l.Warn().Msgf("resolved orphaned worker semaphore slots for tenant %s", tenantIdStr)
+		}
+
+		if err != nil {
+			t.l.Err(err).Msgf("could not resolve orphaned worker semaphore slots for tenant %s", tenantIdStr)
```
Comment (abelanger5) [inline internal/services/ticker/ticker.go:389]: Return from error?

File: internal/services/ticker/ticker.go [hunk 3]
```diff
@@ -350,21 +367,72 @@ func (t *TickerImpl) runStreamEventCleanup(ctx context.Context) func() {
...
+		default:
+		}
+
+		n, err := t.repo.Worker().ResolveWorkerSemaphoreSlots(ctx, tenantId)
+
+		if n.HasResolved {
+			t.l.Warn().Msgf("resolved orphaned worker semaphore slots for tenant %s", tenantIdStr)
+		}
+
+		if err != nil {
```
Comment (abelanger5) [inline internal/services/ticker/ticker.go:388]: Handle `err` before looking at `n.HasResolved`?

File: internal/services/ticker/ticker.go [hunk 4]
```diff
@@ -350,21 +367,72 @@ func (t *TickerImpl) runStreamEventCleanup(ctx context.Context) func() {
...
+		if !n.HasMore {
+			return nil
+		}
+	}
+}
+
 func (t *TickerImpl) runWorkerSemaphoreSlotResolver(ctx context.Context) func() {
 	return func() {
-		ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+		// we set requeueAfter to 4 seconds in the future to avoid requeuing the same step run multiple times
```
Comment (abelanger5) [inline internal/services/ticker/ticker.go]: Change comment copypasta

**Review by abelanger5 — CHANGES_REQUESTED** [review]

File: internal/services/ticker/ticker.go (hunk 1 above)
Comment (grutt) [reply internal/services/ticker/ticker.go:131]:
```
func WithPartitionId(pid string) TickerOpt {
	return func(opts *TickerOpts) {
		opts.partitionId = pid
	}
}
```

I am

Comment (abelanger5) [reply internal/services/ticker/ticker.go:131]: That's in the `TickerOpts`, but unless I'm missing something it's not set in the TickerImpl

**Review by abelanger5 — APPROVED** [review]

**Pull request merged**